Bringing the UC2 controllers and widgets to the main branch

beniroquai opened the pull request:
This is a huge pull request and I honestly don't know how to deal with it. Perhaps we have a session to work on it? @jacopoabramo ?

beniroquai commented:
I think the first task would be to resolve the test-cases I guess? ;)

beniroquai commented:
@jacopoabramo I have adapted the tests slightly. The test_scan is broken it seems..

beniroquai commented:
@jacopoabramo help :D 
https://github.com/openUC2/ImSwitch/actions/runs/6304767094/job/17127008001#step:6:159

All tests completed, but a cryptic error. Any ideas why?

jacopoabramo commented:
Checking the macos-latest unit test report:

>       self._esp32 = uc2.UC2Client(host=self._host, port=80, identity=self._identity, serialport=self._serialport, baudrate=115200, logger=self.__logger)
E       TypeError: __init__() got an unexpected keyword argument 'logger'

This seems to be the first reason that the unit test fails; I'm pretty sure new errors will pop up slowly but surely

beniroquai commented:
This issue has been resolved already. So there is a new error saying:

```
imswitch/imcontrol/_test/unit/test_recording.py: 197 warnings
  /Users/runner/work/ImSwitch/ImSwitch/imswitch/imcontrol/model/managers/RecordingManager.py:537: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    self.__recordingManager.sigRecordingTimeUpdated.emit(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================= 20 passed, 218 warnings in 84.28s (0:01:24) ==================
Error: Process completed with exit code 139.
```

jacopoabramo commented:
Interesting, it seems that the runner crashes but I could not for the life of me understand why this happens. I'll try to check what that error code means

jacopoabramo commented:
@beniroquai please replace all the calls to `print` with logger calls. You can create a local reference to the central logger instance with the `initLogger` function in `imswitch.imcommon.model`

jacopoabramo commented:
I tested the changes locally and run the unit test; they seem to pass. It would be great though if you could add some unit test of your own to increase coverage, @beniroquai

beniroquai commented:
Thanks for commenting! I cleaned up some unused controllers and replaced prints with logs. 
I had one question. In my last commit [4448ace](https://github.com/ImSwitch/ImSwitch/pull/186/commits/4448acebedf96615e741c1edbd625d9792ce8481) I reverted a change which overwrote the `tryInherit`-stuff. My impression was that this slows down ImSwitch's startup. Any comments on that?

jacopoabramo commented:
The initLogger function tries to inherit from a parent of the object that is creating the logger instance, so to log referring to the parent and not the current object.

From the function:

```py
    logger = None

    if tryInheritParent:
        # Use logger from first parent in stack that has one
        for frameInfo in inspect.stack():
            frameLocals = frameInfo[0].f_locals
            if 'self' not in frameLocals:
                continue

            parent = frameLocals['self']
            parentRef = weakref.ref(parent)
            if parentRef not in objLoggers:
                continue

            logger = objLoggers[parentRef]
            break
```

If there's a deep stack of objects it may take a while.

jacopoabramo commented:
@beniroquai can you run the the failing test locally to see what's the output?

beniroquai commented:
On my mac I get this:

```bash
(imswitch) (base) MacBook-Pro-9:ImSwitch bene$ python -m pytest --pyargs imswitch -v --ignore=imswitch/imcontrol/_test/ui/test_example_setups.py
/Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/paramiko/transport.py:219: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
================================================================ test session starts =================================================================
platform darwin -- Python 3.9.15, pytest-7.2.1, pluggy-1.0.0 -- /Users/bene/opt/anaconda3/envs/imswitch/bin/python
cachedir: .pytest_cache
PyQt5 5.15.9 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2
rootdir: /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch
plugins: qt-4.2.0, napari-plugin-engine-0.2.0, anyio-3.7.1, npe2-0.7.2, napari-0.4.18
collected 20 items                                                                                                                                   

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error httpd started on port 5001
INFO:     Started server process [56111]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
PASSED                                                                    [  5%]
imswitch/imcontrol/_test/ui/test_liveview.py::test_grid_no_error PASSED                                                                        [ 10%]
imswitch/imcontrol/_test/ui/test_liveview.py::test_crosshair_no_error PASSED                                                                   [ 15%]
imswitch/imcontrol/_test/ui/test_liveview.py::test_close_no_error PASSED                                                                       [ 20%]
imswitch/imcontrol/_test/ui/test_without_widgets.py::test_without_widgets_starts_no_error httpd failed to start on port 5001
[Errno 48] Address already in use
INFO:     Started server process [56111]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
ERROR:    [Errno 48] error while attempting to bind on address ('0.0.0.0', 8000): address already in use
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
PASSED                                               [ 25%]
imswitch/imcontrol/_test/ui/test_without_widgets.py::test_without_widgets_close_no_error PASSED                                                [ 30%]
imswitch/imcontrol/_test/unit/test_acquisition.py::test_acquisition_liveview_single[detectorInfos0] PASSED                                     [ 35%]
imswitch/imcontrol/_test/unit/test_acquisition.py::test_acquisition_liveview_single[detectorInfos1] PASSED                                     [ 40%]
imswitch/imcontrol/_test/unit/test_acquisition.py::test_acquisition_liveview_multi[Camera 1] PASSED                                            [ 45%]
imswitch/imcontrol/_test/unit/test_acquisition.py::test_acquisition_liveview_multi[Camera 2] PASSED                                            [ 50%]
imswitch/imcontrol/_test/unit/test_recording.py::test_recording_spec_frames[detectorInfos0-10] PASSED                                          [ 55%]
imswitch/imcontrol/_test/unit/test_recording.py::test_recording_spec_frames[detectorInfos1-53] PASSED                                          [ 60%]
imswitch/imcontrol/_test/unit/test_recording.py::test_recording_spec_time[detectorInfos0] PASSED                                               [ 65%]
imswitch/imcontrol/_test/unit/test_recording.py::test_recording_spec_time[detectorInfos1] PASSED                                               [ 70%]
imswitch/imcontrol/_test/unit/test_recording.py::test_recording_spec_time[detectorInfos2] PASSED                                               [ 75%]
imswitch/imcontrol/_test/unit/test_scan.py::test_scan_signals PASSED                                                                           [ 80%]
imswitch/imcontrol/_test/unit/test_stores.py::test_storer_instatiation PASSED                                                                  [ 85%]
imswitch/imcontrol/_test/unit/test_stores.py::test_zarr_storer PASSED                                                                          [ 90%]
imswitch/imcontrol/_test/unit/test_stores.py::test_tiff_storer PASSED                                                                          [ 95%]
imswitch/imcontrol/_test/unit/test_stores.py::test_hdf5_storer PASSED                                                                          [100%]

================================================================== warnings summary ==================================================================
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/google_crc32c/__config__.py:18
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/google_crc32c/__config__.py:18: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.cloud')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2350
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2350
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2350: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(parent)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165: RuntimeWarning: Missing value of non-optional type magnefication detected when decoding DPCInfo and was defaulted to None by infer_missing=True. Set infer_missing=False (the default) to prevent this behavior.
    warnings.warn(

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165: RuntimeWarning: Missing value of non-optional type swapImageAxes detected when decoding FocusLockInfo and was defaulted to None by infer_missing=True. Set infer_missing=False (the default) to prevent this behavior.
    warnings.warn(

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165: RuntimeWarning: Missing value of non-optional type piKp detected when decoding FocusLockInfo and was defaulted to None by infer_missing=True. Set infer_missing=False (the default) to prevent this behavior.
    warnings.warn(

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165: RuntimeWarning: Missing value of non-optional type piKi detected when decoding FocusLockInfo and was defaulted to None by infer_missing=True. Set infer_missing=False (the default) to prevent this behavior.
    warnings.warn(

../../../../../../../opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/dataclasses_json/core.py:165: RuntimeWarning: Missing value of non-optional type scanWidgetType detected when decoding ScanInfo and was defaulted to None by infer_missing=True. Set infer_missing=False (the default) to prevent this behavior.
    warnings.warn(

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PYTHON/nanoimagingpack/NanoImagingPack/coordinates.py:9: DeprecationWarning: Please use `gaussian_filter` from the `scipy.ndimage` namespace, the `scipy.ndimage.filters` namespace is deprecated.
    from scipy.ndimage.filters import gaussian_filter

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PYTHON/nanoimagingpack/NanoImagingPack/image.py:24: PendingDeprecationWarning: Importing from numpy.matlib is deprecated since 1.19.0. The matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray. 
    from numpy.matlib import repmat

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PYTHON/nanoimagingpack/NanoImagingPack/image.py:32: DeprecationWarning: Please use `center_of_mass` from the `scipy.ndimage` namespace, the `scipy.ndimage.measurements` namespace is deprecated.
    from scipy.ndimage.measurements import center_of_mass as cm

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/aionotify/aioutils.py:123: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def stream_from_fd(fd, loop):

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/jupyter_client/connect.py:20: DeprecationWarning: Jupyter is migrating its paths to use standard platformdirs
  given by the platformdirs library.  To remove this warning and
  see the appropriate new directories, set the environment variable
  `JUPYTER_PLATFORM_DIRS=1` and then run `jupyter --paths`.
  The use of platformdirs will be the default in `jupyter_core` v6
    from jupyter_core.paths import jupyter_data_dir, jupyter_runtime_dir, secure_write

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcommon/view/guitools/naparitools.py:99: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    napariViewer.window.qt_viewer.dockLayerList

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcommon/view/guitools/naparitools.py:102: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    napariViewer.window.qt_viewer.dockLayerList.qt_area,

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcommon/view/guitools/naparitools.py:103: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    napariViewer.window.qt_viewer.dockLayerList

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcommon/view/guitools/naparitools.py:105: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    napariViewer.window.qt_viewer.dockLayerList.show()

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcontrol/view/widgets/ImageWidget.py:110: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    canvas=self.napariViewer.window.qt_viewer.canvas,

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcontrol/view/widgets/ImageWidget.py:111: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    view=self.napariViewer.window.qt_viewer.view,

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcontrol/view/widgets/ImageWidget.py:112: FutureWarning: Public access to Window.qt_viewer is deprecated and will be removed in
  v0.5.0. It is considered an "implementation detail" of the napari
  application, not part of the napari viewer model. If your use case
  requires access to qt_viewer, please open an issue to discuss.
    parent=self.napariViewer.window.qt_viewer.view.scene,

imswitch/imcontrol/_test/ui/test_liveview.py: 18 warnings
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/colour/utilities/array.py:648: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    return dtype(a)  # type: ignore[arg-type, return-value]

imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
imswitch/imcontrol/_test/ui/test_liveview.py::test_liveview_no_error
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/pyqtgraph/parametertree/Parameter.py:136: DeprecationWarning: Using "values" to set limits is deprecated. Use "limits" instead.
    return cls(**opts)

imswitch/imcontrol/_test/unit/test_recording.py: 130 warnings
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/h5py/_hl/filters.py:250: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.
    chunks = guess_chunk(shape, maxshape, dtype.itemsize)

imswitch/imcontrol/_test/unit/test_recording.py: 84 warnings
imswitch/imcontrol/_test/unit/test_stores.py: 6 warnings
  /Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/h5py/_hl/attrs.py:103: DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.
    self.create(name, data=value)

imswitch/imcontrol/_test/unit/test_recording.py: 80 warnings
  /Users/bene/Dropbox/Dokumente/Promotion/PROJECTS/MicronController/PYTHON/ImSwitch/imswitch/imcontrol/model/managers/RecordingManager.py:537: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    self.__recordingManager.sigRecordingTimeUpdated.emit(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================== 20 passed, 359 warnings in 193.99s (0:03:13) ====================================================
/Users/bene/opt/anaconda3/envs/imswitch/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py:293: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  warnings.warn(
Segmentation fault: 11
(imswitch) (base) MacBook-Pro-9:ImSwitch bene$ 
```

a quick search suggests that the Segmentation fault maybe an issue with tensorflow/pytorch..however, this is not even imported anymore. I don't know what this means.
Source: https://github.com/magenta/magenta/issues/1801


What kind of testcases would you like to see in the PR to be added

jacopoabramo commented:
Well, that's funky. Were you importing one of those packages before? If so where? I'll take a look and review it (it's easier to ask you since the PR is massive).

As for test cases, would be nice to have a mock configuration of a standard UC2 setup and try to imitate the examples in the `_test/unit` folder.

jacopoabramo commented:
```
imswitch/imcontrol/_test/ui/test_without_widgets.py::test_without_widgets_starts_no_error httpd failed to start on port 5001
[Errno 48] Address already in use
INFO:     Started server process [56111]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
ERROR:    [Errno 48] error while attempting to bind on address ('0.0.0.0', 8000): address already in use
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
PASSED  
```

Could be nice if a local server to attach ImSwitch to in the unit tests could be added as well... Like a mock server or something

jacopoabramo commented:
@beniroquai I think you should probably update this branch with the changes from the master branch. In order:

- merge from ImSwitch:master to ImSwitch:uc2
- merge from ImSwitch:uc2 to openUC2:master

Hopefully we should sync up.

beniroquai commented:
@jacopoabramo, we should definitely think about a versioning system. E.g. CLI when uploading to pypi.

beniroquai commented:
What is your favorite way to automate this?

jacopoabramo commented:
I'm not experienced enough to setup a workflow for automatically packaging to PyPi when a new version is released but [here](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/) they seem to describe the process. If you have a preferred way I'm all hears, this is new stuff to me

beniroquai commented:
It has already been established, but replacing the version number is missing. So either we do a PR on new version and replace the __version__ string with the version TAG or we use this as the version TAG . the workflow is inside github actions. the uc2 branch works

jacopoabramo commented:
I'm sorry, I got lost. Could you clarify this for me?

beniroquai commented:
First thing: I merged the imswitch/master into imswitch/uc2. Did that help? ;) 
However, the PR I filed is trying to merge openuc2/master into imswitch/uc2..so what's next?

The workflow is here:https://github.com/openUC2/ImSwitch/blob/master/.github/workflows/imswitch-pypi.yml#L30 with replacement of the version-number

jacopoabramo commented:
> The workflow is here:https://github.com/openUC2/ImSwitch/blob/master/.github/workflows/imswitch-pypi.yml#L30 with replacement of the version-number

Ohh nice, this is perfect. And this is part of the PR?

> However, the PR I filed is trying to merge openuc2/master into imswitch/uc2..so what's next?

I checked and now since your `ImSwitch:uc2` branch is up to date with the master, it should be fine. I don't know why I was thinking otherwise... now we just need to fix these failing test

beniroquai commented:
@jacopoabramo, quick question. Are these changes based on the Imswitch/UC2 or the openUC2/master branch? I guess some of the changes I already made .. but in the openUC2 repo

jacopoabramo commented:
@beniroquai some of the changes were already existing in the master repo, exactly. If they are already there you can just mark them as done.

beniroquai commented on imswitch/imblockly/controller/ImScrMainController.py:
```diff
@@ -0,0 +1,125 @@
+        self.server_thread = None
+
+    class StaticServer(BaseHTTPRequestHandler):
+        def do_GET(self):
```
The current version looks as follows:

```py
    class StaticServer(BaseHTTPRequestHandler):
        def do_GET(self):
            root = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "view", "static", "index.html")
            if self.path == '/':
                filename = root + '/index.html'
            else:
                filename = root + self.path
            self.send_response(200)
            if filename[-4:] == '.css':
                self.send_header('Content-type', 'text/css')
            elif filename[-5:] == '.json':
                self.send_header('Content-type', 'application/javascript')
            elif filename[-3:] == '.js':
                self.send_header('Content-type', 'application/javascript')
            elif filename[-4:] == '.ico':
                self.send_header('Content-type', 'image/x-icon')
            else:
                self.send_header('Content-type', 'text/html')
            self.end_headers()
            with open(filename, 'rb') as fh:
                html = fh.read()
                self.wfile.write(html)
```

beniroquai commented on .github/workflows/imswitch-packconda.yml:
Good question. I'll remove it for now. I attempted to have imswitch built in the Conda world. But too many non-conda dependencies avoid this effort.

beniroquai commented on README.md:
```diff
@@ -41,8 +41,8 @@ You will then be able to start ImSwitch with this command:
 
+(Developers installing ImSwitch from the source repository should run `pip install -r requirements-dev.txt` instead, and start it using ``python -m imswitch``)
 
 ### Option C: Install from Github (UC2 version)
```
removed.

beniroquai commented on construct.yaml:
artifact to package imswitch into conda. removed

beniroquai commented on docs/changelog.rst:
```diff
@@ -2,6 +2,20 @@
+
+- Added support for Pulse Streamer and other devices, and improve low lever managers: https://github.com/kasasxav/ImSwitch/pull/1
+- Server-client support with fastAPI and Pyro5: https://github.com/kasasxav/ImSwitch/tree/master/imswitch/imcontrol/controller/server
+- Added features for open microscopy hardware (OpenUC2, SQUID, ESP32) more on beniroquai's fork
```
I removed the changelog

beniroquai commented on docs/conf.py:
```diff
@@ -12,7 +12,7 @@
 
 project = 'ImSwitch'
-copyright = '2020-2021 ImSwitch developers'
+copyright = '2020-2022 ImSwitch developers'
```
Done

beniroquai commented on imswitch/_data/user_defaults/uc2_configs/pindefUC2Standalon1.json:
I rather remove it.

beniroquai commented on imswitch/app/public/assets/css/style.css:
It was a test to control ImSwitch from Typescript. I remove it.

beniroquai commented on imswitch/app/public/assets/js/index.js:
nope. Removed it

beniroquai commented on imswitch/imblockly/__init__.py:
No, it's very experimental and breaking since the QT Webview is not really supported anymore (?)

beniroquai commented on imswitch/imblockly/controller/CommunicationChannel.py:
Sure, who could write this? :-P
I would rather state: "experimental" and then if there is somebody keen on playing with it, fine. This could be a `first nice plugin` ;-)

beniroquai commented on imswitch/imblockly/controller/ImScrMainViewController.py:
```diff
@@ -0,0 +1,45 @@
+from .ConsoleController import ConsoleController
+from .EditorController import EditorController
+#from .BlockController import BlockController
```
Done

beniroquai commented on imswitch/imblockly/controller/ImScrMainViewController.py:
```diff
@@ -0,0 +1,45 @@
+from .basecontrollers import ImScrWidgetController
+
+
+class ImScrMainViewController(ImScrWidgetController):
```
As said, the imblockly was experimental. Maybe I create a branch and leave it there?

beniroquai commented on imswitch/imblockly/model/ScriptExecutor.py:
Ok, I'll remove imblockly once and for all

beniroquai commented on imswitch/imcommon/view/guitools/joystick.py:
Would you mind doing that? ; )

beniroquai commented on imswitch/imcontrol/__test_Manager.py:
I guess this is not coming from me.

beniroquai commented on imswitch/imblockly/controller/ImScrMainController.py:
```diff
@@ -0,0 +1,128 @@
+
+
+
+class MyServer:
```
It's gone now.

beniroquai commented on imswitch/imcommon/model/api.py:
Nope, streaming images and retreiving images via HTTP

beniroquai commented on imswitch/imcontrol/_test/__init__.py:
I haven't touched it - i guess ;)

beniroquai reviewed (commented):
Hey @jacopoabramo I have removed ImBlockly for now as it's too confusing. I have updated many parts of the repo according to your suggestions. What do you think?

beniroquai commented:
@jacopoabramo should I change the target branch to become dev or uc2?

jacopoabramo commented:
@beniroquai there are some conflicts to resolve first. I'll take a look locally

beniroquai commented:
Thanks @jacopoabramo. Let me know what I have to do know ;)

jacopoabramo commented:
I managed to solve some of the conflicts. You should add your updates on that PR side

beniroquai commented:
Aren't they automatically reflected? I see all the latest changes. So should be good to go! :)

jacopoabramo commented:
Oh really? Cool then
